homeworks/module2/hw2/task2/main.go

```diff
@@ -0,0 +1,508 @@
+package main
+
+import "log"
+
+type Node struct {
+	size   int
+	key    [3]int
+	first  *Node
+	second *Node
+	third  *Node
+	fourth *Node
+	parent *Node
+}
+
+func NewNode(k int) *Node {
+	return &Node{
+		size:   1,
+		key:    [3]int{k, 0, 0},
+		first:  nil,
+		second: nil,
+		third:  nil,
+		fourth: nil,
+		parent: nil,
+	}
+}
+
+func (n *Node) swap(x, y *int) {
+	*x, *y = *y, *x
+}
+
+func (n *Node) insertToNode(k int) {
+	n.key[n.size] = k
+	n.size++
+
+	if n.size == 1 {
+		return
+	}
+
+	if n.size == 2 {
+		if n.key[0] > n.key[1] {
+			n.swap(&n.key[0], &n.key[1])
+		}
+	}
+
+	if n.size == 3 {
+		if n.key[0] > n.key[1] {
+			n.swap(&n.key[0], &n.key[1])
+		}
+		if n.key[0] > n.key[2] {
+			n.swap(&n.key[0], &n.key[2])
+		}
+		if n.key[1] > n.key[2] {
+			n.swap(&n.key[1], &n.key[2])
+		}
+	}
+}
+
+func (n *Node) removeFromNode(k int) {
+	if n.size >= 1 && n.key[0] == k {
+		n.key[0] = n.key[1]
+		n.key[1] = n.key[2]
+		n.size--
+	} else if n.size == 2 && n.key[1] == k {
+		n.key[1] = n.key[2]
+		n.size--
+	}
+}
+
+func (n *Node) Insert(k int) *Node {
+	if n == nil {
+		return NewNode(k)
+	}
+
+	if n.first == nil && n.second == nil && n.third == nil {
+		n.insertToNode(k)
+	} else if k <= n.key[0] {
+		n.first = n.first.Insert(k)
+	} else if n.size == 1 || (n.size == 2 && k <= n.key[1]) {
+		n.second = n.second.Insert(k)
+	} else {
+		n.third = n.third.Insert(k)
+	}
+
+	return n.Split(n)
+}
+
+func (n *Node) Split(item *Node) *Node {
+	if item.size < 3 {
+		return item
+	}
+
+	x := NewNode(item.key[0])
+	y := NewNode(item.key[2])
+
+	x.first = item.first
+	x.second = item.second
+	y.first = item.third
+	y.second = item.fourth
+
```

Comment:
вот тут еще предка забыл добавить для х и у 
```suggestion
	x.parent = item.parent
	y.parent = item.parent
```
из-за этого дерево после добавления седьмого элемента получается не валидным (нарушается принцип "Все листовые вершины находятся на одном уровне (на нижнем уровне) и содержат 1 или 2 поля.")